kernel/src/hil/digest.rs

```diff
@@ -0,0 +1,48 @@
+//! Interface for Digest
+
+use crate::returncode::ReturnCode;
+
+/// Implement this trait and use `set_client()` in order to receive callbacks.
+pub trait Client<'a> {
+    /// On error `result` will contain device specific error codes
+    fn hash_done(&'a self, result: Result<(), u32>);
+}
+
+/// Computes a digest (cryptographic hash) over data
+pub trait Digest<'a> {
+    /// Set the client instance which will receive `hash_done()` callbacks.
+    /// This callback is called when a digest is computed.
+    /// The callback should follow the `Client` `hash_done` callback.
+    fn set_client(&'a self, client: &'a dyn Client<'a>);
+
+    /// Add data to the digest block. This is the data that will be used
+    /// for the hash function.
+    /// Returns the number of bytes written on success
+    /// The data can be fed to the underlying hardware block either syncronously
+    /// or asyncronously. This is up to the hardware implementation and is
+    /// transperent to the user.
+    fn add_data(&self, data: &[u8]) -> Result<usize, ReturnCode>;
+
+    /// Request the hardware block to generate a Digest
+    /// This doesn't return anything, instead the client needs to have
+    /// set a `hash_done` handler to determine when this is complete.
+    /// If there is data from the `add_data()` command asyncrously waiting to
+    /// be written it will be written before the operation starts.
+    fn run(&'a self);
+
+    /// Stores the returned digest in the memory location specified. This
+    /// is the only way to get the output of the digest.
+    fn get_digest(&self, digest: &mut [u8]) -> ReturnCode;
```

Comment:
This is unnecessary. You should instead pass a slice in the `hash_done` callback. Having this method only introduces more code paths that could be erroneous: a client can call `get_digest()` even when the digest is not ready. Also a client would be able to call `get_digest()` `n`-times, requiring the implementer to handle these scenarios. It's much easier if the generated digest is given once in `hash_done` - which given a correct implementation is guaranteed to be only called as a result of a successful operation, and only once.

Reply:
The only problem then is that we need some call that specifies where to save the digest. I guess the best option is to add an argument to `run()` and then remove this function. I can do that.

Reply:
That is an option, though not my favorite. I get why you can't hand out a simple reference to an `u8` slice in the callback - internally you're working with `u32`s. But bear in mind that passing the buffer in the callback does not require any persistent allocation. It is enough to pass in a reference to the buffer on the call stack. As that isn't too big, I think it should be preferred here. The client would need to copy the contents in the callback function anyways.

A viable alternative could be to pass a `&mut dyn Iterator<type = u8>` to the callback. That way you never have to allocate the full buffer and can build the `u8` iterator from your `u32`s on the fly. However that introduces some level of complexity - it may be easier to have a stack allocated temporary buffer as proposed above.